File: backend/api/db/models/report_file.py
from sqlalchemy import Enum, Text, DateTime, ForeignKey, func
from sqlalchemy.orm import Mapped, mapped_column, relationship
from datetime import datetime
import enum
from .report import Base


class FileType(str, enum.Enum):
    pdf = "pdf"
    json = "json"
    txt = "txt"


class FileCategory(str, enum.Enum):
    source = "source"
    extract = "extract"
    output = "output"


class FileStatus(str, enum.Enum):
    pending = "pending"
    processing = "processing"
    done = "done"
    error = "error"


class ReportFile(Base):
    __tablename__ = "report_files"

    id: Mapped[int] = mapped_column(primary_key=True)
    report_id: Mapped[int] = mapped_column(ForeignKey("reports.id", ondelete="CASCADE"))
    type: Mapped[FileType] = mapped_column(Enum(FileType, name="file_type"))
    category: Mapped[FileCategory] = mapped_column(
        Enum(FileCategory, name="file_category")
    )
    status: Mapped[FileStatus] = mapped_column(
        Enum(FileStatus, name="file_status"), default=FileStatus.pending
    )
    s3_bucket: Mapped[str] = mapped_column(Text)
    s3_key: Mapped[str] = mapped_column(Text)
    error: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    report = relationship("Report", back_populates="files")

File: frontend/src/types/report-file.ts
export const FileCategory = {
  source: "source",
  extract: "extract",
  output: "output",
} as const;
export type FileCategory = (typeof FileCategory)[keyof typeof FileCategory];

export const FileType = {
  pdf: "pdf",
  json: "json",
  txt: "txt",
} as const;
export type FileType = (typeof FileType)[keyof typeof FileType];

export const FileStatus = {
  pending: "pending",
  processing: "processing",
  done: "done",
  error: "error",
} as const;
export type FileStatus = (typeof FileStatus)[keyof typeof FileStatus];

export interface ReportFile {
  id: number;
  report_id: number;
  type: FileType;
  category: FileCategory;
  status: FileStatus;
  s3_bucket: string;
  s3_key: string;
  error?: string;
  created_at: string;
  updated_at: string;
}
